Makefile

```diff
@@ -0,0 +1,116 @@
+BUILD_DIR=build
+TIDB_SOURCE=$(BUILD_DIR)/tidb
+TIKV_SOURCE=$(BUILD_DIR)/tikv
+PD_SOURCE=$(BUILD_DIR)/pd
+ARTIFACT_BINARY=$(BUILD_DIR)/bin
+ARTIFACT_DIR=$(BUILD_DIR)/dist
+ARTIFACT_DOCKER=${ARTIFACT_DIR}/tidb-docker.tar.gz
+ARTIFACT_PACKAGE=$(ARTIFACT_DIR)/tidb-pkg
+ARTIFACT_RPM=${ARTIFACT_DIR}/
+DOCKER_IMAGE_NAME=pingcap/tidb:$(TAG)
+
+define fetch_source
+	mkdir -p $(1)
+	@if [ -f $(1)/Makefile ]; then\
+		cd $(1) && git fetch; \
+	else \
+		git clone https://github.com/pingcap/$(2).git $(1); \
+	fi
+endef
+
+$(BUILD_DIR):
+	mkdir -p $(BUILD_DIR)
+
+# don't use directory so we can force update the projects each time.
+TIDB_SOURCE:
```

Comment:
Tasks without a target path should have `.PHONY: TIDB_SOURCE` which tells make that if a file called `./TIDB_SOURCE` exists it should still do this task.